src/launcher/index.ts

```diff
@@ -0,0 +1,226 @@
+/**
+ * The launcher token.
+ */
+export const ILauncher = new Token<ILauncher>('jupyter.services.launcher');
+/* tslint:enable */
+
+
+/**
+ * The class name added to LauncherWidget instances.
+ */
+const LAUNCHER_CLASS = 'jp-LauncherWidget';
+
+/**
+ * The class name added to LauncherWidget image nodes.
+ */
+const IMAGE_CLASS = 'jp-LauncherWidget-image';
+
+/**
+ * The class name added to LauncherWidget text nodes.
+ */
+const TEXT_CLASS = 'jp-LauncherWidget-text';
+
+/**
+ * The class name added to LauncherWidget column nodes.
+ */
+const COLUMN_CLASS = 'jp-LauncherWidget-column';
+
+/**
+ * The class name added to LauncherWidget folder node.
+ */
+const FOLDER_CLASS = 'jp-LauncherWidget-folder';
+
+/**
+ * The class name added to LauncherWidget path nodes.
+ */
+const PATH_CLASS = 'jp-LauncherWidget-path';
+
+/**
+ * The class name added to LauncherWidget current working directory node.
+ */
+const CWD_CLASS = 'jp-LauncherWidget-cwd';
+
+/**
+ * The class name added to LauncherWidget body nodes.
+ */
+const BODY_CLASS = 'jp-LauncherWidget-body';
+
+/**
+ * The class name added to LauncherWidget dialog node.
+ */
+const DIALOG_CLASS = 'jp-LauncherWidget-dialog';
+
+
+/**
+ * The launcher interface.
+ */
+export interface ILauncher {
+  /**
+   * Add a command item to the Launcher, and trigger re-render event for parent
+   * widget.
+   *
+   * @param name - The display name.
+   *
+   * @param action - The command that should be executed on clicking.
+   *
+   * @param args - Arguments to the `action` command.
+   *
+   * @param imgName - The CSS class to attach to the item. Defaults to
+   * 'jp-Image' followed by the `name` with spaces removed. So if the name is
+   * 'Launch New Terminal' the class name will be 'jp-ImageLaunchNewTerminal'.
+   *
+   * @returns A disposable that will remove the item from Launcher, and trigger
+   * re-render event for parent widget.
+   *
+   */
+  add(name: string, action: string, args?: JSONObject, imgName?: string): IDisposable ;
+}
+
+
+/**
+ * Simple encapsulation of name and callback of launcher entries. This is an
+ * implementation detail of the LauncherModel. You should not need to use
+ * this class directly, but use the `LauncherModel.add` method instead.
+ */
+export class LauncherItem {
+
+  readonly name: string;
+
+  readonly clickCallback: () => void;
+
+  readonly imgName: string;
+
+  /**
+   * Construct a new launcher item.
+   */
+  constructor(name: string, clickCallback: () => void, imgName?: string) {
+    this.name = name;
+    this.clickCallback = clickCallback;
+    if (imgName) {
+      this.imgName = imgName;
+    } else {
+      this.imgName = 'jp-Image' + name.replace(/\ /g, '');
+    }
+  }
+}
+
+
+/**
+ * LauncherModel keeps track of the path to working directory and has a list of
+ * LauncherItems, which the LauncherWidget will render.
+ */
+export class LauncherModel extends VDomModel implements ILauncher {
+  items: LauncherItem[] = [];
+  _path: string = 'home';
```

Comment:
From https://github.com/jupyterlab/jupyterlab/wiki/TypeScript-Style-Guide#names, this should be a private property, and from https://github.com/jupyterlab/jupyterlab/wiki/TypeScript-Style-Guide#classes, it should be at the end of the definition.